plugins/stub_resolver/stub_resolver_tests.ml

```diff
@@ -0,0 +1,219 @@
+  let syms =
+    List.fold symbols ~init:(Map.empty (module String))
+      ~f:(fun syms (name,data) ->
+          Map.add_exn syms name data) in
+  let prog = create_program syms in
+  let expected =
+    List.fold expected
+      ~init:(Map.empty (module Tid))
+      ~f:(fun tids (stub, impl) ->
+          Map.add_exn tids
+            (tid_for_name_exn prog stub)
+            (tid_for_name_exn prog impl)) in
+  let pairs = Stub_resolver.run prog in
+  assert_bool name (Map.equal Tid.equal expected pairs)
+
+let symbol ?(is_stub=false) ?(aliases=[]) name =
+  name, {is_stub; aliases}
+
+let test name ~symbols ~expected =
+  name >:: run name symbols expected
+
+let suite = "stub-resolver" >::: [
+
+    test "simple case: we have pairs"
+      ~symbols:[
+        symbol "alpha";
+        symbol "alpha_2" ~is_stub:true ~aliases:["alpha"];
+      ]
+      ~expected:["alpha_2", "alpha"];
+
+    test "simple case: no pairs"
+      ~symbols:[
+        symbol "bravo";
+        symbol "bravo_2" ~is_stub:true;
+      ]
+      ~expected:[];
+
+    test "simple case: still no pairs"
+      ~symbols:[
+        symbol "charlie";
+        symbol "charlie_2" ~is_stub:true ~aliases:["some-sym"];
+      ]
+      ~expected:[];
+
+    test "stubs only"
+      ~symbols:[
+        symbol "delta" ~is_stub:true;
+        symbol "delta_2" ~is_stub:true ~aliases:["delta"];
+      ]
+      ~expected:[];
+
+    test "impl only"
+      ~symbols:[
+        symbol "echo";
+        symbol "echo_2" ~aliases:["echo"];
+      ]
+      ~expected:[];
+
+    test "impl can be aliases as well"
+      ~symbols:[
+        symbol "foxtrot" ~aliases:["golf"];
+        symbol "golf" ~is_stub:true;
+      ]
+      ~expected:["golf", "foxtrot"];
+
+    test "many alias"
+      ~symbols:[
+        symbol "hotel" ~aliases:["india"; "juliet"];
+        symbol "india" ~is_stub:true;
+      ]
+      ~expected:["india", "hotel"];
+
+    test "ambiguous impl"
+      ~symbols:[
+        symbol "kilo" ~aliases:["kilo_2"; "kilo_3"];
+        symbol "kilo_2" ~is_stub:true;
+        symbol "kilo_3" ~is_stub:true;
+      ]
+      ~expected:[];
+
+    test "ambiguous stubs"
+      ~symbols:[
+        symbol "lima";
+        symbol "lima_2" ~is_stub:true ~aliases:["lima"];
+        symbol "lima_3" ~is_stub:true ~aliases:["lima"];
+      ]
+      ~expected:[];
+
+    test "crossreference"
+      ~symbols:[
+        symbol "mike" ~aliases:["november"];
+        symbol "november" ~is_stub:true ~aliases:["mike"];
+      ]
+      ~expected:["november", "mike"];
+
+    test "many pairs"
+      ~symbols:[
+        symbol "papa" ;
+        symbol "quebec";
+        symbol "romeo";
+        symbol "siera" ~is_stub:true ~aliases:["papa"];
+        symbol "tango" ~is_stub:true ~aliases:["quebec"];
+        symbol "uniform" ~is_stub:true ~aliases:["romeo"];
+
+      ]
+      ~expected:["siera",   "papa";
+                 "tango",   "quebec";
+                 "uniform", "romeo";];
```

Comment:
We need cases where symbols have many aliases, some intersecting aliases and some not, but where there is more than two cases, e.g.,

```
test "several intersections" [
  stub, "a",  ["a1"; "a2"; "a3"];
  real, "a0", ["a"];
  stub, "b",  ["b1"; "b2"; "b3"];
  real, "b0", ["b"; "c"];
  real, "c",  ["d"];
  stub, "d",  [];
] ~expect:[
  "a", "a0";
]
```

and more tests like this.